File: src/main/java/io/nshusa/rsam/binary/sprite/ImageArchive.java
package io.nshusa.rsam.binary.sprite;

import io.nshusa.rsam.binary.Archive;
import io.nshusa.rsam.util.HashUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public final class ImageArchive {

    private int hash;
    private final List<Sprite> sprites = new ArrayList<>();

    public ImageArchive(int hash) {
        this.hash = hash;
    }

    public static ImageArchive decode(Archive archive, int hash) {
        ImageArchive imageArchive = new ImageArchive(hash);

        for (int i = 0; ; i++) {
            try {
                Sprite decoded = Sprite.decode(archive, hash, i);

                imageArchive.sprites.add(decoded);
            } catch (IOException e) {
                break;
            }
        }

        return imageArchive;
    }

    public static ImageArchive decode(Archive archive, String name) {
        return decode(archive, HashUtils.nameToHash(name));
    }

    public int getHash() {
        return hash;
    }

    public void setHash(int hash) {
        this.hash = hash;
    }

    public void setName(String name) {
        this.hash = HashUtils.nameToHash(name);
    }

    public List<Sprite> getSprites() {
        return sprites;
    }

}

File: src/main/kotlin/io/nshusa/rsam/binary/sprite/ImageArchive.kt
package io.nshusa.rsam.binary.sprite

import io.nshusa.rsam.binary.Archive
import io.nshusa.rsam.util.HashUtils

import java.io.IOException
import java.util.ArrayList

class ImageArchive(var hash: Int) {
    private val sprites = ArrayList<Sprite>()

    fun setName(name: String) {
        this.hash = HashUtils.nameToHash(name)
    }

    fun getSprites(): List<Sprite> {
        return sprites
    }

    companion object {

        fun decode(archive: Archive, hash: Int): ImageArchive {
            val imageArchive = ImageArchive(hash)

            var i = 0
            while (true) {
                try {
                    val decoded = Sprite.decode(archive, hash = hash, id = i)

                    imageArchive.sprites.add(decoded)
                } catch (e: IOException) {
                    break
                }

                i++
            }

            return imageArchive
        }

        fun decode(archive: Archive, name: String): ImageArchive {
            return decode(archive, HashUtils.nameToHash(name))
        }
    }

}